2014-01-24  Richard Henderson <rth@redhat.com>

	* sysdeps/alpha/tls-macros.h (TLS_GD): Add dependency on $gp.
	(TLS_LD, TLS_IE): Likewise.

diff --git a/ports/sysdeps/alpha/tls-macros.h b/ports/sysdeps/alpha/tls-macros.h
index 0385d93..00489c2 100644
--- a/ports/sysdeps/alpha/tls-macros.h
+++ b/ports/sysdeps/alpha/tls-macros.h
@@ -2,21 +2,21 @@
 
 extern void *__tls_get_addr (void *);
 
-# define TLS_GD(x)						\
-  ({ void *__result;						\
-     asm ("lda %0, " #x "($gp) !tlsgd" : "=r" (__result));	\
+# define TLS_GD(x)							\
+  ({ register void *__gp asm ("$29"); void *__result;			\
+     asm ("lda %0, " #x "($gp) !tlsgd" : "=r" (__result) : "r"(__gp));	\
      __tls_get_addr (__result); })
 
-# define TLS_LD(x)						\
-  ({ void *__result;						\
-     asm ("lda %0, " #x "($gp) !tlsldm" : "=r" (__result));	\
-     __result = __tls_get_addr (__result);			\
-     asm ("lda %0, " #x "(%0) !dtprel" : "+r" (__result));	\
+# define TLS_LD(x)							\
+  ({ register void *__gp asm ("$29"); void *__result;			\
+     asm ("lda %0, " #x "($gp) !tlsldm" : "=r" (__result) : "r"(__gp));	\
+     __result = __tls_get_addr (__result);				\
+     asm ("lda %0, " #x "(%0) !dtprel" : "+r" (__result));		\
      __result; })
 
-# define TLS_IE(x)						\
-  ({ long ofs;							\
-     asm ("ldq %0, " #x "($gp) !gottprel" : "=r"(ofs));		\
+# define TLS_IE(x)							\
+  ({ register void *__gp asm ("$29"); long ofs;				\
+     asm ("ldq %0, " #x "($gp) !gottprel" : "=r"(ofs) : "r"(__gp));	\
      __builtin_thread_pointer () + ofs; })
 
 # define TLS_LE(x)						\
